### PR TITLE
Adds ribbon in starter screen if mon has beat classic mode

### PR DESCRIPTION
**Pull request by shayebeadling**
Heard lots of people liking the idea of rewarding classic runs so this is a super simple implementation that could get the ball rolling;
1. Adds a winCount variable in the starterDataEntry object to keep track of classic wins on a mon.
2. Adds a red ribbon that shows on an entry if the winCount > 0.

I implore anyone that is more artistically inclined to make a better ribbon if this idea is adopted, cause mine looks a crewmate from among us.

image

UPDATED: Moved the count from dexData to starterData. Incrementation works, but it does seem finicky like Akuma said. Gonna step through and see what the deal is.

UPDATED AGAIN: Incrementation is now working consistently.

### CODE REVIEW

**Comment by Akuma-Reiki** [comment]
Good launching point!
Would suggest something like 
```for (const pokemon of this.scene.getParty()) {
              const speciesId = getPokemonSpecies(pokemon.species.speciesId)
              this.scene.gameData.incrementStarterWinCount(speciesId);
            }```
in phases.ts line 3498

and then something like
```incrementStarterWinCount(species: PokemonSpecies):void {
  const speciesWinCount = this.dexData[species.speciesId].winCount;

    if(Number.isNaN(speciesWinCount))
      this.dexData[species.speciesId].winCount=0;
    
    this.dexData[species.speciesId].winCount++;
    console.log(this.dexData[species.speciesId].winCount);
  }```
  
  at line 1081 of game-data.ts
  
  Played around with those a bit and could make it show the ribbon upon beating classic mode, but it didn't apply it to the last pokemon
  I think the way around this is by making it async / a promise
  
  
Also think that winCount would be better as starterData rather than dexData, so if you win with ivysaur, you get the ribbon on bulbasaur

**Review by Akuma-Reiki — CHANGES_REQUESTED** [review]
1083 of game-data.ts seems to cause a freeze / crash if the pokemon is evolved.
Perhaps some way of finding the species ID of the prevolution of the pokemon similar to how candy does it?

```
incrementStarterWinCount(species: PokemonSpecies): void {
    const hasPrevolution = pokemonPrevolutions.hasOwnProperty(species.speciesId);
    if (!hasPrevolution){
      if (!this.starterData[species.speciesId].winCount) 
        this.starterData[species.speciesId].winCount = 0;
      this.starterData[species.speciesId].winCount++;
    }
    
   //logic for getting the prevolution species ID and then updating the wincount for it

  }
```

**Comment by Akuma-Reiki** [comment]
working with Brain Frog to get a better version of this, but this has the functionality

**Review by Akuma-Reiki — CHANGES_REQUESTED** [review]
```
incrementStarterWinCount(species: PokemonSpecies): void {
    const hasPrevolution = pokemonPrevolutions.hasOwnProperty(species.speciesId);
    const starterIndex = hasPrevolution ? getPokemonSpecies(pokemonPrevolutions[species.speciesId]).speciesId : species.speciesId ;

    if (!this.starterData[starterIndex].winCount) 
        this.starterData[starterIndex].winCount = 0;
    this.starterData[starterIndex].winCount++;
  }
```

Me and Brain Frog got this working, should work with the current call of the function in phases.ts

only direction to go from here is maybe changing the champion ribbon icon

**Review by Akuma-Reiki — APPROVED** [review]
lgtm !

File: src/phases.ts [hunk 1]
```diff
@@ -3499,6 +3499,10 @@ export class GameOverPhase extends BattlePhase {
           if (this.scene.gameMode.isClassic) {
             firstClear = this.scene.validateAchv(achvs.CLASSIC_VICTORY);
             this.scene.gameData.gameStats.sessionsWon++;
+            for (let pokemon of this.scene.getParty()) {
```
Comment (TempsRay) [inline src/phases.ts:3502]: Maybe here could be where the 5 egg voucher per new ribbon awarded can go? Something like this?

```ts
for (let pokemon of this.scene.getParty()) {
  const rootSpeciesId = pokemon.getRootSpeciesId(true);
  if (this.scene.gameData.starterData[rootSpeciesId] == 0) {
    this.scene.pushPhase(new ModifierRewardPhase(this.scene, modifierTypes.VOUCHER_PLUS));
  }
}
```

**Comment by shayebeadling** [comment]
OK the way that this behaves now is winning in classic with a mon will reward them, and any prevolution starters of them with a ribbon. If it is their first win, it displays a phase in the postgame that rewards you with a 5-pull voucher

**Comment by bennybroseph** [comment]
Looks good on my end. Tested all the combos I could think. I always got the expected tickets including the final 10x for beating classic in general.

Two starters who have never beaten classic. Expected 2 5x tickets

https://github.com/pagefaultgames/pokerogue/assets/13838608/49d6d0cf-bd61-47d2-9e99-39da25a5620b

Two starters who have never beaten classic and one that evolved. Expected 3 5x tickets

https://github.com/pagefaultgames/pokerogue/assets/13838608/205618fc-a58a-4eda-8bfc-ca6f66ad0aba

No starters who have beaten classic but one that evolved. Expected 1 5x tickets 

https://github.com/pagefaultgames/pokerogue/assets/13838608/67553422-8304-4d36-8b5e-bf71402f71e2

Six starters who have never beaten classic. Expected 6 5x tickets

https://github.com/pagefaultgames/pokerogue/assets/13838608/c49f390e-e334-4ec8-ae04-659d3337f303

One starter that has never beater classic and one that has but has now evolved. Expected 1 5x tickets

https://github.com/pagefaultgames/pokerogue/assets/13838608/6926b0d9-bf2c-4632-8165-8ca28287b778

All starters have beaten classic but they are different forms/evolutions. Expected 0 5x tickets

https://github.com/pagefaultgames/pokerogue/assets/13838608/c7dffdc8-9fda-4e74-a316-98dcd5b552db

One starter who has never beaten classic and a freshly caught pokemon that has never beaten classic. Expected 2 5x tickets

https://github.com/pagefaultgames/pokerogue/assets/13838608/2d769218-783b-4edf-a58b-0d3f9e3b496d

**Review by TempsRay — APPROVED** [review]
The next big version update!

**Review by Flashfyre — APPROVED** [review]

**Pull request merged**